sips/sip-011/sip-011-catastrophic-failure-recovery.md

```diff
@@ -0,0 +1,763 @@
+sign a Stacks node software release that implements the soft fork rules.  This
+is the first step in this process because this new node software must address
+whatever catastrophic bug that led to the need to follow this procedure.  The
+remainder of this procedure is focused on codifying the new rules via the SIP
+process in order to require them in future node software releases.  The act of
+releasing this new node software will follow the critical bugfix point-release
+procedure, but with an additional warning from the announcement email that this
+new release is the beginning of a soft-fork roll-out.
+2. Once the catastrophic bug has been addressed, and a reliable majority of
+blocks on the canonical fork are produced by miners running the new software, a
+SIP describing the soft fork will be written and published describing the new
+rules and why they are necessary.  This SIP must then be advanced in the SIP
+process to "activation-in-progress" status by the relevant advisory boards and
+technical steering committee.  Among other things, this SIP's activation
+criteria must require the following:
+
+   * At least 80% of all mined blocks must _explicitly and continuously vote to
+     accept_ the soft fork rules for at least _two consecutive_ whole PoX reward
+cycle prepare phases. Additional, stricter criteria are permitted.
+   * An _activation window_, defined by a _start activation height_ and an _end
+     activation height_, represented as burnchain block heights, must be
+stipulated in the activation criteria.  The aforementioned miner vote must take
+place and run to completion in-between these two block heights.  The block
+heights _must_ correspond to the beginnings of PoX prepare phases, and must be
+between 14 reward cycles (1 maximum Stacking period plus two reward cycles) and
+25 reward cycles (about 1 year) apart.
+   * A soft-fork-specific bit pattern must be specificied, which miners _must_
+     include in the unused `memo` bits in their block-commit transactions on the
+burnchain in order to unambiguously signal support for activating the new rules.
+Different soft forks may (re)use the same bit pattern only if they have
+non-overlapping activation windows.
+
+3. Once a release with the soft fork voting rules and activation window is
+written and adequately tested (possibly including rolling it out to one or more
+public testnets first), it will be announced via a signed email from
+announce@stacks.org.  This release announcement must occur before the activation
+window begins.
+4. If at least 80% of all sortitions in two consecutive, whole PoX prepare
+   phases vote to accept the soft fork rules within the activation window, then the new
+rules will take effect starting in the next whole reward cycle (i.e. right after
+the second prepare phase finishes).  All new releases of the Stacks node will
+adhere to the soft fork rules, since they are now part of the block validation rules.
```

Comment:
In this soft-fork process, would the node released in step 1 _immediately_ begin enforcing the soft-fork rules (i.e., ignoring blocks that do not follow the soft fork)? Then is step 4 just a consolidation of this rule (i.e., the declaration that all future releases will also follow the soft fork rule)?

Reply:
Not necessarily.  This SIP only requires that miners who vote for the soft-fork do so through this signaling procedure, and that the new rules must take effect _no later than_ when this activation threshold is met.  The SIP intentionally does not specify anything about how early miners can start deciding when to ignore blocks they would no longer consider valid.  In fact, per point 2, the provision that "stricter criteria are permitted" is meant to allow a particular soft-fork upgrade to impose additional requirements for miners to begin activating the new rules.  For example, in the third case study below, a particular soft fork to repair a bug in smart contract processing could require miners to begin orphaning blocks in which the bug manifests ASAP.